### PR TITLE
Add `optuna.visualization.plot_edf` function.

### PR DESCRIPTION
**Pull request by sile**
## Motivation

This PR aims to provide a new way to analyze search spaces. EDF (empirical distribution function) is a useful tool for such a purpose. Please refer to [Designing Network Design Spaces](https://arxiv.org/abs/2003.13678) for the detail.

## Description of the changes

Add `optuna.visualization.plot_edf` function which can plot the objective value EDF of a study (or multiple studies). 

## Rendered Doc

image
image

### CODE REVIEW

File: optuna/visualization/_edf.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        trials = [
+            trial
+            for trial in study.get_trials(deepcopy=False)
+            if trial.state == TrialState.COMPLETE
+        ]
+
+        y_values = []
+        for x in x_values:
+            y = len([trial for trial in trials if trial.value <= x]) / len(trials)
+            y_values.append(y)
```
Comment (hvy) [inline optuna/visualization/_edf.py:131]: Somewhat simplify-able.
```suggestion
        values = np.asarray(
            [
                trial.value
                for trial in study.get_trials(deepcopy=False)
                if trial.state == TrialState.COMPLETE
            ]
        )

        y_values = np.sum(values[:, np.newaxis] <= x_values, axis=0) / values.size
```

File: optuna/visualization/_edf.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+        raise ValueError("No studies were given.")
+
+    layout = go.Layout(
+        title="EDF Plot",
+        xaxis={"title": "Objective Value"},
+        yaxis={"title": "Cumulative Probability"},
+    )
+
+    all_trials = list(
+        itertools.chain.from_iterable(
```
Comment (hvy) [inline optuna/visualization/_edf.py:102]: First time stumbling upon `from_iterable`, great to know. 👍

File: optuna/visualization/_edf.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+    return _get_edf_plot(studies)
+
+
+def _get_edf_plot(studies: List[Study]) -> "go.Figure":
+    if len(studies) == 0:
+        raise ValueError("No studies were given.")
+
+    layout = go.Layout(
+        title="EDF Plot",
```
Comment (hvy) [inline optuna/visualization/_edf.py:96]: Maybe it'd be more user friendly spelling "EDF" out. What do you think?

File: optuna/visualization/_edf.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        studies = study
+    else:
+        studies = [study]
+
+    return _get_edf_plot(studies)
+
+
+def _get_edf_plot(studies: List[Study]) -> "go.Figure":
+    if len(studies) == 0:
+        raise ValueError("No studies were given.")
```
Comment (HideakiImamura) [inline optuna/visualization/_edf.py:93]: Why don't you return the empty `go.Figure` in this case?

File: optuna/visualization/_edf.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        xaxis={"title": "Objective Value"},
+        yaxis={"title": "Cumulative Probability"},
+    )
+
+    all_trials = list(
+        itertools.chain.from_iterable(
+            (
+                trial
+                for trial in study.get_trials(deepcopy=False)
+                if trial.state == TrialState.COMPLETE
```
Comment (HideakiImamura) [inline optuna/visualization/_edf.py:106]: I think we should explain the completed trials are only considered for visualization in the documentation.

File: optuna/visualization/_edf.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+from optuna.trial import TrialState
+from optuna.visualization._plotly_imports import _imports
+
+if _imports.is_successful():
+    from optuna.visualization._plotly_imports import go
+
+_logger = get_logger(__name__)
+
+
+def plot_edf(study: Union[Study, List[Study]]) -> "go.Figure":
```
[the remark this answers is not in the document]
Comment (sile) [reply optuna/visualization/_edf.py:18]: Sounds good. Thanks!

File: optuna/visualization/_edf.py (hunk 3 above)
Comment (sile) [reply optuna/visualization/_edf.py:97]: Sorry but please let me confirm your intention.
Did you suggest a change like the following?
```suggestion
        title="Empirical Distribution Function Plot",
```
If so, I think it's a good idea.

File: optuna/visualization/_edf.py (hunk 4 above)
Comment (sile) [reply optuna/visualization/_edf.py:94]: Honestly, there is no reason we should raise an error. I'll rewrite the code to return the `go.Figure` as you suggested. Thanks!

File: optuna/visualization/_edf.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sile) [reply optuna/visualization/_edf.py]: Fixed: https://github.com/optuna/optuna/pull/1482/commits/d307d193c277a840fbbfdb73c217fe0ffe9857e4

File: optuna/visualization/_edf.py (hunk 1 above)
Comment (sile) [reply optuna/visualization/_edf.py:132]: Cool!

File: optuna/visualization/_edf.py (hunk 4 above)
Comment (sile) [reply optuna/visualization/_edf.py]: Fixed: https://github.com/optuna/optuna/pull/1482/commits/c5def17727d6153606741ffaf4461c767ea478bc

File: optuna/visualization/_edf.py (hunk 5 above)
Comment (sile) [reply optuna/visualization/_edf.py:110]: Added the description: https://github.com/optuna/optuna/pull/1482/commits/f600f7b2f80ac2f9d82267198fe02e774927f67e

**Comment by sile** [comment]
@HideakiImamura 

> I think we should document that the described EDF curve is based on the uniform distribution for the objective value. If we introduce the other distributions for the objective, we can add other types of EDF curve.

Thank you for your comment but please let me clarify it as I'm not sure that I understand your proposal correctly.
Did you mean we should add a note comment as follows?
image

File: optuna/visualization/_edf.py (hunk 3 above)
Comment (hvy) [reply optuna/visualization/_edf.py:96]: Yes, precisely. Could you go ahead apply it?

File: optuna/visualization/_edf.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    """Plot the objective value EDF (empirical distribution function) of a study.
+
+    Note that only the complete trials are considered when plotting the EDF.
+
+    .. note::
+
+        EDF is useful to analyze and improve search spaces.
+        For instance, you can see a practical use case of EDF in the paper
+        `Designing Network Design Spaces <ttps://arxiv.org/abs/2003.13678>`_.
+
```
Comment (HideakiImamura) [inline optuna/visualization/_edf.py:29]: @sile Thanks for the reply. I think we can note the underlying distribution for the objective in the documentation as follows.
```suggestion

    .. note::

        The plotted EDF assumes that the value of the objective function is in 
        accordance with the uniform distribution over the objective space.
        
```

Comment (sile) [reply optuna/visualization/_edf.py:29]: Thanks!

File: optuna/visualization/_edf.py (hunk 3 above)
Comment (sile) [reply optuna/visualization/_edf.py]: Sure! Applied the above suggestion. Thanks!

**Comment by sile** [comment]
# Re-rendered Doc

image
image

**Pull request merged**